typesafeconfig/typesafe_config.go

```diff
@@ -0,0 +1,340 @@
+	"errors"
+	"fmt"
+	"github.com/armory-io/go-commons/secrets"
+	"github.com/mitchellh/mapstructure"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+	"golang.org/x/exp/maps"
+	"gopkg.in/yaml.v3"
+	"io/fs"
+	"os"
+	"os/user"
+	"path/filepath"
+	"reflect"
+	"strings"
+)
+
+var NoConfigurationSourcesProvided = errors.New("no configuration sources provide, you must provided at least 1 embed.FS or dir path")
+
+type resolver struct {
+	log                 *zap.SugaredLogger
+	embeddedFilesystems []*embed.FS
+	configurationDirs   []string
+	baseNames           []string
+	profiles            []string
+}
+
+type Option = func(resolver *resolver)
+
+func EmbeddedFilesystems(embeddedFilesystems ...*embed.FS) Option {
+	return func(resolver *resolver) {
+		resolver.embeddedFilesystems = embeddedFilesystems
+	}
+}
+
+func Directories(directories ...string) Option {
+	return func(resolver *resolver) {
+		resolver.configurationDirs = directories
+	}
+}
+
+func ActiveProfiles(profiles ...string) Option {
+	return func(resolver *resolver) {
+		resolver.profiles = profiles
+	}
+}
+
+func BaseConfigurationNames(baseNames ...string) Option {
+	return func(resolver *resolver) {
+		resolver.baseNames = baseNames
+	}
+}
+
+func defaultResolver() *resolver {
+	configurationDirs := []string{"/opt/go-application/config"}
+	usr, err := user.Current()
+	if err == nil {
+		configurationDirs = append(configurationDirs, filepath.Join(usr.HomeDir, ".armory"))
+	}
+
+	profiles := strings.Split(os.Getenv("PROFILES_ACTIVE"), ",")
+
+	return &resolver{
+		baseNames:         []string{"application"},
+		configurationDirs: configurationDirs,
+		profiles:          profiles,
+	}
+}
+
+// Resolve given the provided options resolves your configuration
+func Resolve[T any](log *zap.SugaredLogger, options ...Option) (*T, error) {
+	r := defaultResolver()
+	for _, option := range options {
+		option(r)
+	}
+
+	if len(r.embeddedFilesystems) == 0 && len(r.configurationDirs) == 0 {
+		return nil, NoConfigurationSourcesProvided
+	}
+
+	candidates := getConfigurationFileCandidates(r.configurationDirs, r.baseNames, r.profiles)
+	sources, err := loadFileBasedConfigurationSources(log, candidates, r.embeddedFilesystems)
+	if err != nil {
+		return nil, err
+	}
+	sources = append(sources, loadEnvironmentSources())
+	untypedConfig := mergeSources(sources...)
+	if err = resolveSecrets(untypedConfig); err != nil {
+		return nil, err
+	}
+	var typeSafeConfig *T
+	decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		WeaklyTypedInput: true,
+		Result:           &typeSafeConfig,
+		MatchName: func(mapKey, fieldName string) bool {
+			normalizedMapKey := strings.ToLower(mapKey)
+			normalizedMapKey = strings.ReplaceAll(normalizedMapKey, "-", "")
+			normalizedMapKey = strings.ReplaceAll(normalizedMapKey, "_", "")
+			return strings.ToLower(fieldName) == normalizedMapKey
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	err = decoder.Decode(untypedConfig)
+	return typeSafeConfig, err
```

Comment:
I guess this could be `return typeSafeConfig, decoder.Decode(untypedConfig)` but that's probably more personal style